kratos/future/solving_strategies/builders/block_builder.h

```diff
@@ -0,0 +1,375 @@
+        TSystemVectorType& rRHS) const
+    {
+        // Set the free DOFs vector (0 means fixed / 1 means free)
+        // Note that we initialize to 1 so we start assuming all free
+        // Also note that the type is uint_8 for the sake of efficiency
+        const std::size_t system_size = rLHS.size1();
+        std::vector<uint8_t> free_dofs_vector(system_size, 1);
+
+        // Loop the DOFs to find which ones are fixed
+        // Note that DOFs are assumed to be numbered consecutively in the block building
+        const auto dof_begin = rDofArray.begin();
+        IndexPartition<std::size_t>(rDofArray.size()).for_each([&](IndexType Index){
+            const auto p_dof = dof_begin + Index;
+            if (p_dof->IsFixed()) {
+                free_dofs_vector[p_dof->EffectiveEquationId()] = 0;
+            }
+        });
+
+        //TODO: Implement this in the CSR matrix or here?
+        // // Detect if there is a line of all zeros and set the diagonal to a certain number if this happens (1 if not scale, some norms values otherwise)
+        // mScaleFactor = TSparseSpace::CheckAndCorrectZeroDiagonalValues(rModelPart.GetProcessInfo(), rA, rb, mScalingDiagonal);
+
+        // Get the diagonal scaling factor
+        const double diagonal_value = this->GetDiagonalScalingFactor(rLHS);
+
+        // Apply the free DOFs (i.e., fixity) vector to the system arrays
+        rLHS.ApplyHomogeneousDirichlet(free_dofs_vector, diagonal_value, rRHS);
+    }
+
+    /**
+     * @brief Applies the Dirichlet conditions to the system
+     * This method applies the Dirichlet conditions following a block-type build
+     * @param rDofArray DOFs array from which the fixity is checked
+     * @param rRHS Right Hand Side vector
+     */
+    void ApplyBlockBuildDirichletConditions(
+        const DofsArrayType& rDofArray,
+        TSystemVectorType& rRHS) const
+    {
+        // Loop the DOFs to find which ones are fixed
+        // Note that DOFs are assumed to be numbered consecutively in the block building
+        const auto dof_begin = rDofArray.begin();
+        IndexPartition<std::size_t>(rDofArray.size()).for_each([&](IndexType Index){
+            auto p_dof = dof_begin + Index;
+            if (p_dof->IsFixed()) {
+                rRHS[p_dof->EffectiveEquationId()] = 0;
+            }
+        });
+    }
+
+    /**
+     * @brief Get the Diagonal Scaling Factor
+     * This method calculates the diagonal scaling value to be used in the Dirichlet BCs imposition from the provided LHS
+     * @param rLHS Left Hand Side matrix
+     * @return double Diagonal scaling factor
+     */
+    virtual double GetDiagonalScalingFactor(const TSparseMatrixType& rLHS) const
+    {
+        if (mScalingType == ScalingType::NoScaling) {
+            return 1.0;
+        } else if (mScalingType == ScalingType::NormDiagonal) {
+            return rLHS.NormDiagonal();
+            // return rLHS.NormDiagonal() / rLHS.size1(); //TODO: Decide which one
```

Comment:
Currently is the norm, but I think the size ratio is more correct

Reply:
Fair enough. Maybe we add another `ScalingType::NormDiagonalRatio` and set it as default.

Reply:
Done. Now we have the ratios for both the norm and the max diagonal values.